File: bindings/python/examples/chat_example.py
# SPDX-License-Identifier: Apache-2.0 OR MIT
# Copyright 2026 The OpenHTTPA Foundation (openhttpa.org)

"""
OpenHTTPA Python Binding Example

This script demonstrates how to use the `openhttpa` library to:
1. Establish a cryptographically attested session with a TEE.
2. Send a confidential LLM chat request (high-level API).
3. Send a manual trusted request (low-level API).

Requirements:
- openhttpa package (built via maturin)
- Running OpenHTTPA backend (e.g. via `make up` in demo/multiparty-webapp)

Environment variables:
- OPENHTTPA_SERVER / OPENHTTPA_BACKEND_URL: backend URL (default: http://127.0.0.1:8080)
- OPENHTTPA_CALL_TIMEOUT_SECS: per-call deadline in seconds (default: 30)
"""

import concurrent.futures
import os
import sys

import openhttpa

# The address of our OpenHTTPA backend.
SERVER_URI: str = (
    os.getenv("OPENHTTPA_SERVER")
    or os.getenv("OPENHTTPA_BACKEND_URL")
    or "http://127.0.0.1:8080"
)

# Configurable per-call deadline so CI never hangs indefinitely.
CALL_TIMEOUT_SECS: float = float(os.getenv("OPENHTTPA_CALL_TIMEOUT_SECS", "30"))


def run_with_timeout(fn, timeout_secs: float, label: str):
    """Run *fn* in a thread pool and raise TimeoutError if it exceeds *timeout_secs*."""
    with concurrent.futures.ThreadPoolExecutor(max_workers=1) as executor:
        future = executor.submit(fn)
        try:
            return future.result(timeout=timeout_secs)
        except concurrent.futures.TimeoutError:
            raise TimeoutError(
                f"'{label}' did not complete within {timeout_secs:.0f}s. "
                "Check whether the backend is healthy."
            )


def main() -> None:
    print("=== OpenHTTPA Python Example ===")

    # --- 1. High-Level API: Confidential LLM Chat ---
    # This automatically runs the Attestation Handshake (AtHS) under the hood.
    print(f"\n[1] Connecting to {SERVER_URI} for confidential chat...")
    try:
        def _llm_chat():
            llm = openhttpa.PyConfidentialLlm(SERVER_URI, "llama3")
            messages = [
                ("system", "You are a secure assistant running in a TEE."),
                ("user", "Explain why TEEs are useful for LLMs in 2 sentences."),
            ]
            return llm.chat(messages)

        reply = run_with_timeout(_llm_chat, CALL_TIMEOUT_SECS, "confidential LLM chat")
        print(f"    Sending chat request...")
        print(f"    Assistant Reply: {reply}")

    except TimeoutError as exc:
        print(f"    Timeout: {exc}")
        sys.exit(1)
    except Exception as exc:
        print(f"    Error: {exc}")
        print("    (Is the backend running? Run 'make up' in demo/multiparty-webapp)")
        sys.exit(1)

    # --- 2. Low-Level API: Attestation + Trusted Request ---
    print(f"\n[2] Manual Attestation Handshake (AtHS)...")
    try:
        client = openhttpa.PyOpenHttpaClient(SERVER_URI)

        # This performs the X25519 + ML-KEM-768 hybrid handshake.
        session = run_with_timeout(client.attest_handshake, CALL_TIMEOUT_SECS, "attest_handshake")
        print("    Handshake successful!")
        print(f"    Attestation-Binding (AtB) ID: {session.atb_id}")

        print("\n[3] Sending Trusted Request (/api/echo)...")
        # All data sent via trusted_request is AEAD-encrypted (AES-256-GCM)
        # using the keys derived during the handshake.
        payload = b'{"message": "Hello from Python!"}'

        def _trusted_request():
            return client.trusted_request(session, "POST", "/api/echo", payload)

        response_bytes = run_with_timeout(_trusted_request, CALL_TIMEOUT_SECS, "trusted_request")
        print(f"    Response Body: {response_bytes.decode()}")

    except TimeoutError as exc:
        print(f"    Timeout: {exc}")
        sys.exit(1)
    except Exception as exc:
        print(f"    Error during low-level API: {exc}")
        sys.exit(1)

    print("\nSuccess: Protocol integrity verified across Python/Rust/TEE layers.")


if __name__ == "__main__":
    main()

File: bindings/nodejs/examples/chat_example.js
// SPDX-License-Identifier: Apache-2.0 OR MIT
// Copyright 2026 The OpenHTTPA Foundation (openhttpa.org)

/**
 * OpenHTTPA Node.js Binding Example
 *
 * This example uses the @openhttpa/core package (built via napi-rs) to interact
 * with a secure OpenHTTPA enclave.
 *
 * Each network call is wrapped in a deadline guard to ensure the process never
 * hangs indefinitely when the backend is unreachable or unresponsive.
 */

'use strict';

const { attestHandshake, confidentialChat } = require('../index');

const SERVER_URI =
  process.env.OPENHTTPA_SERVER || process.env.OPENHTTPA_BACKEND_URL || 'http://127.0.0.1:8080';

// Configurable per-call deadline in milliseconds (default: 30 s).
const CALL_TIMEOUT_MS = Number(process.env.OPENHTTPA_CALL_TIMEOUT_MS ?? 30_000);

/**
 * Wraps a Promise with a hard deadline.  Rejects with a TimeoutError if the
 * underlying operation does not settle within `ms` milliseconds.
 *
 * @template T
 * @param {Promise<T>} promise - The promise to race against the deadline.
 * @param {number} ms - Timeout in milliseconds.
 * @param {string} label - Descriptive label used in the error message.
 * @returns {Promise<T>}
 */
function withTimeout(promise, ms, label) {
  const deadline = new Promise((_resolve, reject) => {
    const id = setTimeout(() => {
      clearTimeout(id);
      reject(new Error(`Timed out after ${ms}ms waiting for: ${label}`));
    }, ms);
  });
  return Promise.race([promise, deadline]);
}

async function main() {
  console.log('=== OpenHTTPA Node.js Example ===');

  try {
    // --- 1. Attestation Handshake ---
    console.log(`\n[1] Performing Attestation Handshake (AtHS) with ${SERVER_URI}...`);
    const atbId = await withTimeout(
      attestHandshake(SERVER_URI),
      CALL_TIMEOUT_MS,
      'attestHandshake',
    );
    console.log(`    Handshake success! Attestation-Binding ID: ${atbId}`);

    // --- 2. Confidential LLM Chat ---
    console.log(`\n[2] Sending confidential chat request...`);
    const messages = [
      ['system', 'You are a quantum-safe AI assistant.'],
      ['user', 'What is the cipher suite used by OpenHTTPA?'],
    ];

    // confidentialChat handles the handshake automatically if needed,
    // or reuses existing logic in the high-level client.
    const reply = await withTimeout(
      confidentialChat(SERVER_URI, 'llama3', messages),
      CALL_TIMEOUT_MS,
      'confidentialChat',
    );
    console.log(`    Assistant Reply: ${reply}`);

    console.log('\nSuccess: OpenHTTPA protocol verified via Node.js/NAPI-RS.');
  } catch (err) {
    console.error(`\n[!] Error: ${err.message}`);
    console.log("    Ensure the backend is running: 'make up' in demo/multiparty-webapp");
    process.exit(1);
  }
}

main();
